domains/client/block-preprocessor/src/runtime_api_full.rs

```diff
@@ -43,10 +44,22 @@ where
         ext: &Block::Extrinsic,
     ) -> Result<ExtractedStateRoots<Block>, ApiError> {
         let api = self.client.runtime_api();
-        api.extract_xdm_proof_state_roots(best_hash, ext)
-            .and_then(|maybe_state_roots| {
-                maybe_state_roots.ok_or(ApiError::Application("Empty state roots".into()))
-            })
+        let messenger_api_version = api
+            .api_version::<dyn MessengerApi<Block, NumberFor<Block>>>(best_hash)?
+            .ok_or_else(|| {
+                ApiError::Application(
+                    format!("Could not find `MessengerApi` api for block `{best_hash:?}`.").into(),
+                )
+            })?;
+
+        let maybe_state_roots = extract_xdm_proof_state_roots_with_client(
+            messenger_api_version,
+            &*api,
```

Comment:
If `api` is sufficient to retrieve `messenger_api_version`, why does `extract_xdm_proof_state_roots_with_client` both to be provided explicitly?

Reply:
Yeah. That make sense. And it will also reduce the duplicate code of retrieving the messenger api. I have applied this review suggestion.